[BE] feat: 유저 요청시 로깅 추가, 후보지 산출 후 프롬프트에 적용 구현

dbsdndcks opened the pull request:
# #️⃣ Issue Number
#335 
#348 

## 🕹️ 작업 내용

한 줄 요약 : 로그 추가 및 AI 프롬프트에 후보지 목록 추가

- [x] 비즈니스 로직 시작시 User 의 요청 정보가 로그로 기록되는가 (api url, method, ip, userAgent 정보)
- [x] 출발지로 부터 후보지들을 산출하는가
- [x] 추천 결과 지역이  산출한 후보지들에 속하는가

## 📋 리뷰 포인트

1️⃣오늘 백엔드 크루들에게  질문했던 Optional 관련 로직 처리에 대해  의견을 나눠주세요 
더 좋은 방법이 있다면 공유 부탁드립니다!

```
   public Optional<SubwayStation> findByName(final String name) {
        if ("총신대입구역".equals(name) || "이수역".equals(name)) {
            Optional<SubwayStation> subwayStationEntity = getSubwayStation("총신대입구(이수)역");
            if (subwayStationEntity.isPresent()) {
                return subwayStationEntity;
            }
        }
        Optional<SubwayStation> subwayStation = getSubwayStation(name);
        if (subwayStation.isPresent()) {
            return subwayStation;
        }
        return Optional.empty();
    }
```

2️⃣ Entity 클래스가 추가되면서 중간중간 Domain -> Entity 혹은 Entity -> Domain 반환 로직이 포함되어 있습니다.
실수한 부분은 없는지 확인해 주시면 감사하겠습니다 🥹

3️⃣ 아래 코드는 후보지  산출 로직입니다. 가독성 혹은 개선할 부분이 있다면 의견 공유 부탁드리겠습니다.
```
public List<SubwayStation> generateCandidatePlace(List<SubwayStation> startingStations) {

        List<SubwayStationEntity> stationEntities = startingStations.stream()
                .map(SubwayStationEntity::toFromSubwayStation)
                .toList();

        Coordinate[] coordinateArr = getCoordinates(stationEntities);

        org.springframework.data.geo.Point center = getCenterPoint(coordinateArr);
        Distance distance;
            distance = new Distance(DISTANCE_VALUE, Metrics.KILOMETERS);

        return subwayStationRepository.findByPointNear(center, distance).stream()
                .map(SubwayStationEntity::toDomain)
                .toList();
    }

    private Coordinate[] getCoordinates(List<SubwayStationEntity> stationEntities) {
        return stationEntities.stream()
                .map(place -> {
                    List<Double> coordinates = place.getPoint().getCoordinates();
                    return new Coordinate(coordinates.get(0), coordinates.get(1));
                })
                .toArray(Coordinate[]::new);
    }

    private org.springframework.data.geo.Point getCenterPoint(Coordinate[] coordinateArr) {

        MultiPoint multiPoint = geometryFactory.createMultiPointFromCoords(coordinateArr);

        Point centroid = multiPoint.getCentroid();

        return new org.springframework.data.geo.Point(centroid.getX(), centroid.getY());
    }
```

## 🔮 기타 사항

- 감사합니다.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - 출발지·후보지 정보를 함께 사용해 더 정교한 위치 추천 제공
  - 지하철역 중심 좌표를 활용한 후보지 자동 생성

- Bug Fixes
  - 좌표 도메인 클래스 구문 오류 수정으로 안정성 향상

- Chores
  - /locations, /recommendations 엔드포인트 접근 로그 필터 추가
  - 요청 제한 정책 강화(버킷 용량/리필 축소)로 남용 방지
  - 지리정보 관련 라이브러리 업데이트 및 설정 추가
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dbsdndcks committed:
refactor: Redis 버켓 용량 수정

- 1000 -> 500

dbsdndcks committed:
feat: EndPoint 별 로깅 추가

- UserAgent 정보
- Ip 정보
- Request URL 정보
- Request Method 정보

dbsdndcks committed:
refactor: properties 수정

dbsdndcks committed:
feat: 지역 추천로직 변경 과정

dbsdndcks committed:
fix: dev 브랜치 merge Conflict 수정

dbsdndcks committed:
feat: SubwayStationEntity 클래스 추가

dbsdndcks committed:
feat: GeometryConfig 클래스 구현

- GeometryFactory 빈 생성

dbsdndcks committed:
refactor: 공백 및 라인 수정

dbsdndcks committed:
refactor: GeoJsonPoint 객체 반영, SubwayStation -> Entity 객체 변환,
Jts 의존성 추가

dbsdndcks committed:
refactor: 후보지 산출 메서드 분리

dbsdndcks committed:
feat: AI 프롬프트 추가

- 후보지역 리스트도 같이 제공

dbsdndcks committed:
feat: 후보지 산출 후 AI에게 추천받는 로직 구현

dbsdndcks committed:
refactor: 임포트 추가

dbsdndcks committed:
fix: Merge Conflict 해결

dbsdndcks committed:
fix: Merge Conflict 해결

dbsdndcks committed:
test: 테스트 수정

jbilee commented on backend/src/main/java/com/f12/moitz/application/adapter/SubwayRouteFinderAdapter.java:
```diff
@@ -19,13 +19,15 @@
             @Autowired final SubwayEdgeService subwayEdgeService
     ) {
         this.subwayStationService = subwayStationService;
+        this.subwayEdgeService = subwayEdgeService;
```
[이 커밋](https://github.com/woowacourse-teams/2025-moitz/commit/f6b66a5ffa93bd241ad2e9b55489df79bd0a7037)에서 제거된 코드가 다시 들어갔어요!

dbsdndcks replied:
dev에서 최신내역을 다 반영한건데 알 수가 없네요 ㅠㅠ 더블체크해보겠습니다

jbilee commented on .github/.coderabbit.yaml:
이건 삭제된 파일인데 다시 들어가버린거 같아요!

jbilee commented on backend/build.gradle:
```diff
@@ -34,6 +34,8 @@ dependencies {
 	implementation 'com.bucket4j:bucket4j-core:8.10.1'
 	implementation 'com.bucket4j:bucket4j-redis:8.10.1'
 
+	implementation 'org.locationtech.jts:jts-core:1.20.0'
```
42라인에도 같은 패키지에 대한 선언이 있는데, 하나만 지워주면 될거 같아요!

lifeishiphop commented on .github/.coderabbit.yaml:
```diff
@@ -0,0 +1,11 @@
+language: ko-KR
```
이 파일 없어야 해요!

lifeishiphop commented on backend/build.gradle:
```diff
@@ -34,6 +34,8 @@ dependencies {
 	implementation 'com.bucket4j:bucket4j-redis:8.10.1'
 
+	implementation 'org.locationtech.jts:jts-core:1.20.0'
+
```
40번째 라인 확인해주세용!

lifeishiphop commented on backend/src/main/java/com/f12/moitz/application/RecommendationService.java:
```diff
@@ -120,6 +125,12 @@ public String recommendLocation(final RecommendationRequest request) {
         ).toHexString().toUpperCase();
     }
 
+    private List<String> findPlaceNames(List<SubwayStation> startingPlaces) {
```
여기는 내부적으로 getName을 돌고 있으니 getPlaceNames가 어떨까? 싶어요!

갑자기 생각나서 추가로 한 가지 공유하자면
find를 메서드명으로 사용하자니 도메인 반환, Optional 반환의 경우에 혼돈이 와서

도메인 반환 -> get
Optional 반환 -> find
으로 규칙을 정하고 통일하자는 얘기가 나왔었어요!

jbilee commented on backend/src/main/java/com/f12/moitz/application/RecommendationService.java:
```diff
@@ -120,6 +125,12 @@ public String recommendLocation(final RecommendationRequest request) {
         ).toHexString().toUpperCase();
     }
 
+    private List<String> findPlaceNames(List<SubwayStation> startingPlaces) {
```
`startingPlaces`만 받는 로직은 아니어서, 코드래빗 말처럼 매개변수명을 `places`로 하는건 어떨까요!?

추가로 이미 있는 정보를 따로 빼내는 메서드인만큼 메서드명을 getPlaceNames로 하면 어떨까 싶습니다 😀

dbsdndcks replied:
반영확인했습니다!

lifeishiphop commented on backend/src/main/java/com/f12/moitz/application/SubwayStationService.java:
```diff
@@ -1,50 +1,115 @@
+    private Optional<SubwayStation> getSubwayStation(String stationName) {
+        Optional<SubwayStationEntity> subwayStationEntity = subwayStationRepository.findByName(stationName);
+        return subwayStationEntity.map(stationEntity -> Optional.ofNullable(stationEntity.toDomain())).orElse(null);
     }
```
엇 이거 확인해봐야 할 것 같아요!

lifeishiphop replied:
> findByName 메소드도 더 간결하게 정리할 수 있을 것 같아요!

이건 어떤 걸 의미해?

lifeishiphop commented on backend/src/main/java/com/f12/moitz/application/adapter/SubwayRouteFinderAdapter.java:
```diff
@@ -19,13 +19,15 @@
 
     private final SubwayStationService subwayStationService;
+    private final SubwayEdgeService subwayEdgeService;
     private final SubwayEdges subwayEdges;
 
     public SubwayRouteFinderAdapter(
             @Autowired final SubwayStationService subwayStationService,
             @Autowired final SubwayEdgeService subwayEdgeService
     ) {
         this.subwayStationService = subwayStationService;
+        this.subwayEdgeService = subwayEdgeService;
         this.subwayEdges = subwayEdgeService.getSubwayEdges();
     }
```
이 부분은 아마 Conflict 나면서 이전 커밋으로부터 추가가 된 것 같네요 ㅜ ㅜ

dbsdndcks replied:
Conflict 시 이전 커밋으로부터 추가되는 이유를 설명해줘

lifeishiphop commented on backend/src/main/java/com/f12/moitz/application/port/LocationRecommender.java:
```diff
@@ -5,6 +5,5 @@
 
-    RecommendedLocationsResponse recommendLocations(final List<String> startPlaceNames, final String condition);
-
+    RecommendedLocationsResponse recommendLocations(final List<String> startingPlaces,final List<String> candidatePlace, final String requirement);
```
```suggestion
    RecommendedLocationsResponse recommendLocations(final List<String> startingPlaces, final List<String> candidatePlace, final String requirement);
```

사소하지만 White Space 하나 넣고 갈까요?!

dbsdndcks replied:
white space 확인했습니다 !

jbilee commented on backend/src/main/java/com/f12/moitz/application/SetupService.java:
```diff
@@ -35,7 +35,6 @@ public class SetupService {
     private final PlaceFinder placeFinder;
 
     public void setup() {
-        log.info("SubwayEdges 초기화 시작");
```
이건 빼기로 했던 로깅인지 한번 더 확인해보면 좋을듯 해요!

lifeishiphop commented on backend/src/main/java/com/f12/moitz/common/filter/RateLimitFilter.java:
```diff
@@ -52,6 +52,11 @@ public void doFilter(final ServletRequest request, final ServletResponse respons
                         userAgent,
                         probe.getRemainingTokens()
                 );
+                log.info("Request allowed for user: {} | {}, remaining tokens: {}",
+                        clientIp,
+                        userAgent,
+                        probe.getRemainingTokens()
+                );
```
@coderabbitai 현재 개발 서버의 로깅 전략은 Debug, 운영 서버는 Info야. 운영 서버로 넘어가면 Debug 레벨을 확인할 수 없어. 그러면 어떠한 방향으로 나아가는 게 좋을까?

lifeishiphop replied:
@jbilee 이거 한 번만 같이 봐주세요!

jbilee replied:
동일한 메세지로 DEBUG랑 INFO를 같이 찍고 있는데, 개발환경은 DEBUG 이상이면 다 찍히기 때문에 `log.info()`만 남겨도 되지 않을까 싶어요..!

<img width="1159" height="406" alt="스크린샷 2025-09-25 오후 9 04 51" src="https://github.com/user-attachments/assets/51a9ea6e-a97f-4c6f-8851-c9a08bc331b5" />

jbilee commented on backend/src/main/java/com/f12/moitz/infrastructure/client/odsay/OdsayClient.java:
OdsayClient에는 로직적으로 추가된게 없어서 변경사항을 revert해주면 좋을거 같아요!

jbilee commented on backend/src/main/java/com/f12/moitz/infrastructure/client/odsay/OdsayMultiClient.java:
히스토리 관리 차원에서 불필요한 수정사항은 revert하는건 어떠신지요..!

lifeishiphop reviewed (changes requested):
고생 많았서요 레몬 ~~ 🍋 
사소한 코멘트 몇 가지 있으니 확인해주세용

jbilee reviewed (commented):
몇가지 기존 코드에서 사라진거 찾아봤습니다!

lifeishiphop committed:
Resolve merge conflicts in #395

lifeishiphop committed:
chore: 중복 의존성 제거

lifeishiphop committed:
style: LocationRecommender 컨벤션 통일

lifeishiphop committed:
style: RecommendationService 컨벤션 통일

lifeishiphop committed:
fix: Optional 중첩 구조 수정

lifeishiphop commented on backend/src/main/java/com/f12/moitz/common/filter/EndPointLoggingFilter.java:
```diff
@@ -0,0 +1,32 @@
+
+        final HttpServletRequest httpServletRequest = (HttpServletRequest) servletRequest;
+
+        final String clientIp = httpServletRequest.getHeader("X-Forwarded-For");
+        final String userAgent = httpServletRequest.getHeader("User-Agent");
+        final String uri = httpServletRequest.getRequestURI();
+        final String method = httpServletRequest.getMethod();
+
+        log.info("Request URI: {} | Method: {} | ClientIp: {} | User-Agent: {}", uri, method, clientIp, userAgent);
+        filterChain.doFilter(httpServletRequest, servletResponse);
```
1. getRemoteAddr() 가 어떤 값을 가져오는 메서드이고, 그 값을 가져왔을 때 어떻게 활용할 수 있지?
2. 사내망이나 로컬 환경처럼 프록시가 없으면 말 그대로 IP가 없으니 더이상 분석 안 할 필요는 없는지?

lifeishiphop committed:
style: RouteFinderAdapter 컨벤션 통일

lifeishiphop committed:
chore: 미사용 import문 제거

lifeishiphop commented on backend/src/test/java/com/f12/moitz/application/RecommendationServiceTest.java:
```diff
@@ -86,7 +86,7 @@ void recommendLocation_Success() {
                 new RecommendedLocationResponse("선릉역", "이유1", "설명1"),
                 new RecommendedLocationResponse("삼성역", "이유2", "설명2")));
-        given(locationRecommender.recommendLocations(anyList(), anyString())).willReturn(mockLocationsResponse);
+        given(locationRecommender.recommendLocations(anyList(), anyList(), anyString())).willReturn(mockLocationsResponse);
```
현재 테스트 성공하고 있는데?

lifeishiphop committed:
Resolve merge conflicts in #395

lifeishiphop reviewed (approved):
@jbilee 제가 수정했으니 확인해주세용 ~~ 👍

lifeishiphop committed:
style: GeometryConfig 컨벤션 통일

lifeishiphop committed:
refactor: X-Forwarded-For가 없는 경우 커버

jbilee commented on backend/build.gradle:
```diff
@@ -34,10 +34,11 @@ dependencies {
-	annotationProcessor 'org.projectlombok:lombok'
 
-	implementation 'org.locationtech.jts:jts-core:1.19.0'
+    compileOnly 'org.projectlombok:lombok'
```
이 위에 개행이 추가로 생겼는데 이번 PR에서 미리 지우는게 나을까요??

jbilee commented on backend/src/main/java/com/f12/moitz/application/RecommendationService.java:
```diff
@@ -120,6 +125,12 @@ public String recommendLocation(final RecommendationRequest request) {
+        return stations.stream()
+                .map(Place::getName)
+                .toList();
+    }
```
메서드명이 getPlaceNames고 내부 로직에서 추상화된 Place의 메서드를 호출하고 있어서, 매개변수명도 places로 맞추는거 어떻게 생각하시나요!?

jbilee commented on backend/src/main/java/com/f12/moitz/common/filter/EndPointLoggingFilter.java:
```diff
@@ -0,0 +1,33 @@
+public class EndPointLoggingFilter implements Filter {
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
```
여기 매개변수에도 final 붙여줄까요? 👀

jbilee commented on backend/src/main/java/com/f12/moitz/common/filter/EndPointLoggingFilter.java:
```diff
@@ -0,0 +1,33 @@
+
+        log.info("Request URI: {} | Method: {} | ClientIp: {} | User-Agent: {}", uri, method, clientIp, userAgent);
+        filterChain.doFilter(httpServletRequest, servletResponse);
+
+    }
```
이 사이에 개행이 추가로 들어간거 같아요!

jbilee commented on backend/src/main/java/com/f12/moitz/domain/Point.java:
```diff
@@ -26,5 +26,4 @@ private void validate(final double x, final double y) {
             throw new IllegalArgumentException("Y(위도)는 대한민국 영역 내에 있어야 합니다. 범위: 33 ~ 43, 현재 값: " + y);
         }
     }
-
```
여기에서는 필요한 개행이 사라진걸까요? 👀

jbilee commented on backend/src/main/java/com/f12/moitz/application/port/LocationRecommender.java:
```diff
@@ -5,6 +5,9 @@
-    RecommendedLocationsResponse recommendLocations(final List<String> startPlaceNames, final String condition);
-
+    RecommendedLocationsResponse recommendLocations(
+            List<String> startingPlaces,
+            List<String> candidatePlaces,
+            String requirement
```
여기에도 final 붙여줄까요?

lifeishiphop committed:
refactor: SubwayStationService.findByName 함수형 스타일로 개선

egaeng09 commented:
@coderabbitai review again

jbilee commented on backend/src/main/java/com/f12/moitz/domain/subway/SubwayStationEntity.java:
```diff
@@ -0,0 +1,34 @@
+                        subwayStation.getPoint().getX(),
+                        subwayStation.getPoint().getY()
+                ));
+    }
```
이 메서드명은 `fromSubwayStation`을 의도한거겠죠..!?

static 메서드여서 to가 없어도 의도가 드러날거 같아요! -> SubwayStationEntity.fromSubwayStation

lifeishiphop committed:
chore: 미사용 코드 제거

lifeishiphop committed:
style: final 키워드 추가

lifeishiphop committed:
refactor: 메서드 명 일관화 및 패키지 이동

- Entity는 domain -> infrastructure/persistence
- to/from 중복 사용 및 Domain/ClassName 혼용 로직 변경
- 미사용 클래스 제거
- 개행 추가 (컨벤션 통일)

lifeishiphop committed:
style: White Space 및 final 키워드 추가

jbilee reviewed (commented):
트랙킹할 변경사항이 많아서 한번 끊어갈게요..!

lifeishiphop committed:
Resolve merge conflicts in #395

lifeishiphop committed:
refactor: 매개변수를 메서드명과 통일

lifeishiphop committed:
style: 컨벤션 통일

lifeishiphop committed:
chore: 중복 로그 제거

dbsdndcks commented on backend/src/main/java/com/f12/moitz/common/config/WebMvcConfig.java:
```diff
@@ -33,4 +35,16 @@ public FilterRegistrationBean<RateLimitFilter> rateLimitFilterRegistration() {
         return registration;
     }
 
+    @Bean
+    public FilterRegistrationBean<EndPointLoggingFilter> endPointLoggingFilterFilterRegistration() {
+        final FilterRegistrationBean<EndPointLoggingFilter> registration = new FilterRegistrationBean<>();
+
+        registration.setFilter(endPointLoggingFilter);
+        registration.addUrlPatterns("/locations/*");
+        registration.addUrlPatterns("/recommendations/*");
+        registration.setName("endPointLoggingFilter");
+
+        return registration;
+    }
```
setOrder 의 기능을 설명해줘

dbsdndcks committed:
refactor: Filter 실행 순서 명시
